pandas/_libs/tslibs/timestamps.pyx

```diff
@@ -1581,7 +1581,12 @@ class Timestamp(_Timestamp):
                     "Valid values for the fold argument are None, 0, or 1."
                 )
 
-            if (ts_input is not _no_input and not (
+            # GH#52117 note that currently we end up with the year in the ts_input
+            # variable when using positional arguments. We should fix this.
+            from_positional = (ts_input is not None and
```

Comment:
should the ts_input check be for _no_input?

Reply:
Eventually, yes. We have changed the code since pandas 1.5. [Before] (https://github.com/pandas-dev/pandas/blob/778ab8234059059d77eb5d71442c959f8d15c1e0/pandas/_libs/tslibs/timestamps.pyx#L1572), we were building a datetime where we now only raise:
```python
        if tzinfo is not None:
            # GH#17690 tzinfo must be a datetime.tzinfo object, ensured
            #  by the cython annotation.
            if tz is not None:
                if (is_integer_object(tz)
                    and is_integer_object(ts_input)
                    and is_integer_object(freq)
                ):
                    # GH#31929 e.g. Timestamp(2019, 3, 4, 5, 6, tzinfo=foo)
                    # TODO(GH#45307): this will still be fragile to
                    #  mixed-and-matched positional/keyword arguments
                    ts_input = datetime(
                        ts_input,
                        freq,
                        tz,
                        unit or 0,
                        year or 0,
                        month or 0,
                        day or 0,
                        fold=fold or 0,
                    )
                    nanosecond = hour
                    tz = tzinfo
                    return cls(ts_input, nanosecond=nanosecond, tz=tz)

                raise ValueError('Can provide at most one of tz, tzinfo')

            # User passed tzinfo instead of tz; avoid silently ignoring
            tz, tzinfo = tzinfo, None
```

This is what it we do now instead:
```python
        if tzinfo is not None:
            # GH#17690 tzinfo must be a datetime.tzinfo object, ensured
            #  by the cython annotation.
            if tz is not None:
                raise ValueError("Can provide at most one of tz, tzinfo")

            # User passed tzinfo instead of tz; avoid silently ignoring
            tz, tzinfo = tzinfo, None
```

Let me try to fix this by shifting the args one to the right instead.

Reply:
@mroeschke Thanks for giving that example. Turns out, there is another bug caused by the same anomaly in the data. We currently have on main:

```python
pd.Timestamp(2020, 12, 31, second=1)

> Timestamp('2020-12-31 00:00:00.000001')
```

 This looks like a bug to me. In addition, we are testing for this unexpected behavior [here](https://github.com/pandas-dev/pandas/blob/d86cdb0c230b311036d154312448de87e59a6c8b/pandas/tests/scalar/timestamp/test_constructors.py#L341):
```python
    @pytest.mark.parametrize("kwd", ["nanosecond", "microsecond", "second", "minute"])
    def test_constructor_positional_keyword_mixed_with_tzinfo(self, kwd, request):
        # TODO: if we passed microsecond with a keyword we would mess up
        #  xref GH#45307
        if kwd != "nanosecond":
            # nanosecond is keyword-only as of 2.0, others are not
            mark = pytest.mark.xfail(reason="GH#45307")
            request.node.add_marker(mark)

        kwargs = {kwd: 4}
        ts = Timestamp(2020, 12, 31, tzinfo=timezone.utc, **kwargs)

        td_kwargs = {kwd + "s": 4}
        td = Timedelta(**td_kwargs)
        expected = Timestamp("2020-12-31", tz=timezone.utc) + td
        assert ts == expected
```

My proposed fix makes it so that the fails here no longer happen and that:
```python
pd.Timestamp(2020, 12, 31, second=1)

> Timestamp('2020-12-31 00:00:01')
```